### PR TITLE
Add Cloud Build and Build Trigger template

### PR DESCRIPTION
**Pull request by sourced-praveenc**

### CODE REVIEW

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+imports:
+  - path: templates/cloudbuild/trigger.py
+    name: trigger.py
+
+resources:
+  - name: mybuildtrigger
+    type: trigger.py
+    properties:
+      description: <FIXME:buildtrigger_description>
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: Description does not have to be a `fixme`... `fixme` are for things that cannot be hardcoded only.

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      triggerTemplate:
+        repoSource:
+          projectId: <FIXME:project_id>
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: use `$PROJECT_ID` instead.

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+          repoName: <FIXME:repo_name>
+          branchName: <FIXME:branch_name>
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: use `master` ... every repo has a master afaik.

File: community/cloud-foundation/templates/cloudbuild/tests/integration/cloudbuild.bats [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        delete_config
+        rm -f "${RANDOM_FILE}"
+        # delete the container image from registry
+        gcloud container images delete ${IMAGE_TAG}:latest -q
+    fi
+
+    # Per-test teardown steps.
+}
+
+########## TESTS ##########
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/tests/integration/cloudbuild.bats]: where are the tests for triggers?

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_reposource.yaml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  - name: mycloudbuild
+    type: build.py
+    properties:
+      source:
+        repoSource:
+          projectId: <FIXME:project_id>
+          repoName: <FIXME:repo_name>
+          branchName: <FIXME:branch_name>
+      steps:
+        - name: 'gcr.io/cloud-builders/<FIXME:builder_name>'
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_reposource.yaml]: aren't we creating the build steps here? we should pass in the builder name
also I don't believe we need this amount of `FIXME`s; pre-populate as much as possible with build and push steps. Likely the only `FIXME` you need is is the `dir` and `repoName` for anything that is not targeting something the user should must create prior to running the example we should be able to put a value. Also please call out on the example description the pre-work required (i.e. have a source repo etc )

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  - path: templates/cloudbuild/trigger.py
+    name: trigger.py
+
+resources:
+  - name: mybuildtrigger
+    type: trigger.py
+    properties:
+      description: <FIXME:buildtrigger_description>
+      triggerTemplate:
+        repoSource:
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: I'm getting an error when running this example:
```shell
"{\"ResourceType\":\"gcp-types/cloudbuild-v1:cloudbuild.projects.triggers.create\"\
    ,\"ResourceErrorCode\":\"400\",\"ResourceErrorMessage\":{\"code\":400,\"message\"\
    :\"Invalid JSON payload received. Unknown name \\\"repo_source\\\" at 'trigger.trigger_template':\
    \ Cannot find field.\",\"status\":\"INVALID_ARGUMENT\",\"details\":[{\"@type\"\
    :\"type.googleapis.com/google.rpc.BadRequest\",\"fieldViolations\":[{\"field\"\
    :\"trigger.trigger_template\",\"description\":\"Invalid JSON payload received.\
    \ Unknown name \\\"repo_source\\\" at 'trigger.trigger_template': Cannot find\
    \ field.\"}]}],\"statusMessage\":\"Bad Request\",\"requestPath\":\"https://cloudbuild.googleapis.com/v1/projects/ocsi-courseralab1-guestbook/triggers\"\
    ,\"httpMethod\":\"POST\"}}"
```

File: community/cloud-foundation/templates/cloudbuild/trigger.py.schema [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+      The template describing the types of source changes that trigger a build.
+      Branch and tag names in the trigger templates are interpreted as regular
+      expressions. Any branch or tag change that matches that regular
+      expression triggers a build.
+    properties:
+      projectId:
+        type: string
+        description: |
+          The ID of the project that owns the Cloud Source repository.
+          If omitted, the IF of the project requesting the build is assumed.
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/trigger.py.schema]: Did you mean ID?

File: community/cloud-foundation/templates/cloudbuild/trigger.py.schema [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+              description: |
+                The time limit for executing the build step. If not defined,
+                the step has no time limit, and is allowed to continue to run
+                until it either completes or the build itself times out.
+                Defined in seconds with up to nine fractional digits,
+                terminated by 's'; for example: 3.5s.
+      timeout:
+        type: string
+        description: |
+          The time limit for executing the build step. If not defined, the
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/trigger.py.schema]: Is this overall timeout for entire build? I think this description is incorrect - it's outside of the steps object (the one above it is inside steps object).

File: community/cloud-foundation/templates/cloudbuild/trigger.py.schema [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+
+outputs:
+  properties:
+    - id:
+        type: string
+        description: Unique identifier of the trigger.
+    - createTime:
+        type: string
+        description: | 
+          The time the build was requested, in the RFC3339 UTC Zulu format.
```
Comment (fawix) [inline community/cloud-foundation/templates/cloudbuild/trigger.py.schema]: add `"` around `Zulu`

File: community/cloud-foundation/templates/cloudbuild/tests/integration/cloudbuild.bats (hunk 4 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/tests/integration/cloudbuild.bats]: Hi Fatima,

There seems be to no way to either `describe` or `list` triggers using `gcloud` - or at least I didn't find any, which is why the test for `triggers` were not included in the `bats` file.

All I could find is, [this](https://cloud.google.com/cloud-build/docs/running-builds/automate-builds) and this part

image

But, if you do have any links please do let me know and I can add them.

I'm going through your other comments, should be done hopefully by Monday.
Have a good weekend!

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_reposource.yaml (hunk 5 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_reposource.yaml]: Cleaned this up good. Now the only `FIXME` is the `repo_name`
Fix in commit `b806511`

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml (hunk 1 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: This one is also cleaned .. removed all `FIXME` placeholders
Fix in commit `b806511`

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml (hunk 6 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: Sorry bout this. Not sure how this escaped two reviewers eyes
Fix in commit `b806511`

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml (hunk 3 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: Fix in commit `b806511`

File: community/cloud-foundation/templates/cloudbuild/trigger.py.schema (hunk 8 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/trigger.py.schema]: Fix in commit `b806511`

File: community/cloud-foundation/templates/cloudbuild/trigger.py.schema (hunk 9 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/trigger.py.schema]: Fix in commit `b806511`

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml (hunk 2 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: Removed this field - if not provided, dm uses the current project ID

File: community/cloud-foundation/templates/cloudbuild/trigger.py.schema (hunk 7 above)
Comment (sourced-praveenc) [reply community/cloud-foundation/templates/cloudbuild/trigger.py.schema]: `projectId` is the correct field (this is optional)

File: community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml (hunk 6 above)
Comment (fawix) [reply community/cloud-foundation/templates/cloudbuild/examples/cloudbuild_trigger.yaml]: It's ok :) important thing is we caught it.

File: community/cloud-foundation/templates/cloudbuild/tests/integration/cloudbuild.bats (hunk 4 above)
Comment (fawix) [reply community/cloud-foundation/templates/cloudbuild/tests/integration/cloudbuild.bats]: looks like we need a FR for this; will add to our TODO list and merge it for now. Thanks Praveen! Good job cleaning up this PR.

**Pull request merged**